src/BitStringAddresses/bosefs.jl

```diff
@@ -331,10 +331,57 @@ end
 function hopnextneighbour(b::SingleComponentFockAddress, i)
     src = find_occupied_mode(b, (i + 1) >>> 0x1)
     dst = find_mode(b, mod1(src.mode + ifelse(isodd(i), 1, -1), num_modes(b)))
-
     new_b, val = excitation(b, (dst,), (src,))
     return new_b, val
 end
+"""
+    new_address, product = hopnextneighbour(add, chosen, Val(boundary_condition))
+
+Compute the new address of a hopping event for the Hubbard model. Returns the new
+address and the square root of product of occupation numbers of the involved modes
+multiplied by a term consistent with boundary condition.
+
+The off-diagonals are indexed as follows:
+
+* `(chosen + 1) ÷ 2` selects the hopping site.
+* Even `chosen` indicates a hop to the left.
+* Odd `chosen` indicates a hop to the right.
+* Boundary conditions can be periodic, pi-twisted and hard-wall.
+
+# Boundary conditions
+
+* For π-twisted -> :twisted
+* For hard wall -> :hard_wall
+
+# Example
+
+```jldoctest
+julia> using Rimu.Hamiltonians: hopnextneighbour
+
+julia> hopnextneighbour(BoseFS(1, 0, 1), 3, Val(:twisted))
+(BoseFS{2,3}(2, 0, 0), -1.4142135623730951)
+
+julia> hopnextneighbour(BoseFS(1, 0, 1), 3, Val(:hard_wall))
+(BoseFS{2,3}(2, 0, 0), 0.0)
+```
+"""
+function hopnextneighbour(b::SingleComponentFockAddress, i, ::Val{boundary_condition}) where {boundary_condition}
+    new_b,val=hopnextneighbour(b, i)
+
+    if (find_occupied_mode(b,1).mode==1 && i==2) || 
+        (find_occupied_mode(b,num_occupied_modes(b)).mode == num_modes(b) 
+        && i == (2*num_occupied_modes(b)-1))
```

Comment:
I now realise that my previous comments were not very smart 🥲 
- By calling the two-argument version of `hopnextneighbour` first, you now have to call `find_occupied_mode` again twice, which is a non-trivial operation for `BoseFS` with a `BitString`. I may be faster if we can avoid this.
- The branching code is much less readable (as @mtsch  correctly predicted).

Did you check which version runs faster? Did you optimise to get the fastest running code?

Reply:
I have just realised the same thing and when I have checked the run time:
* old version (2-argument version)  took 0.085868 seconds (76.55 k allocations: 5.107 MiB, 99.94% compilation time) and 
*  new version ( just updated) takes 0.029513 seconds (14.98 k allocations: 1001.422 KiB, 99.83% compilation time)

With significant difference. So,You and @mtsch were correct.